// @scalar/openapi-types/dist/schemas/3.1/unprocessed/security-scheme-object.js
import { SecuritySchemeObjectSchema as SecuritySchemeObjectSchema$1, ApiKeyInValues as ApiKeyInValues$1, ApiKeySchema as ApiKeySchema$1, HttpSchema as HttpSchema$1, MutualTlsSchema as MutualTlsSchema$1, OpenIdConnectSchema as OpenIdConnectSchema$1, OAuthFlowsObjectSchema as OAuthFlowsObjectSchema$1, OAuthFlowObjectSchema as OAuthFlowObjectSchema$1, AuthorizationCodeFlowSchema as AuthorizationCodeFlowSchema$1, ClientCredentialsFlowSchema as ClientCredentialsFlowSchema$1, ImplicitFlowSchema as ImplicitFlowSchema$1, PasswordFlowSchema as PasswordFlowSchema$1 } from '../processed/security-scheme-object.js';

/**
 * Security Scheme Object
 *
 * Defines a security scheme that can be used by the operations.
 *
 * Supported schemes are HTTP authentication, an API key (either as a header, a cookie parameter or as a query
 * parameter), mutual TLS (use of a client certificate), OAuth2's common flows (implicit, password, client credentials
 * and authorization code) as defined in RFC6749, and [[OpenID-Connect-Core]]. Please note that as of 2020, the implicit
 * flow is about to be deprecated by OAuth 2.0 Security Best Current Practice. Recommended for most use cases is
 * Authorization Code Grant flow with PKCE.
 *
 * @see https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.1.md#security-scheme-object
 */
const SecuritySchemeObjectSchema = SecuritySchemeObjectSchema$1;
const ApiKeyInValues = ApiKeyInValues$1;
const ApiKeySchema = ApiKeySchema$1;
const HttpSchema = HttpSchema$1;
const MutualTlsSchema = MutualTlsSchema$1;
const OpenIdConnectSchema = OpenIdConnectSchema$1;
const OAuthFlowsObjectSchema = OAuthFlowsObjectSchema$1;
const OAuthFlowObjectSchema = OAuthFlowObjectSchema$1;
const AuthorizationCodeFlowSchema = AuthorizationCodeFlowSchema$1;
const ClientCredentialsFlowSchema = ClientCredentialsFlowSchema$1;
const ImplicitFlowSchema = ImplicitFlowSchema$1;
const PasswordFlowSchema = PasswordFlowSchema$1;

export { ApiKeyInValues, ApiKeySchema, AuthorizationCodeFlowSchema, ClientCredentialsFlowSchema, HttpSchema, ImplicitFlowSchema, MutualTlsSchema, OAuthFlowObjectSchema, OAuthFlowsObjectSchema, OpenIdConnectSchema, PasswordFlowSchema, SecuritySchemeObjectSchema };
